documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+# Summary
+A new API for generic maths in Silk.NET.Maths 3.X, leveraging modern .NET features such as `INumber<T>` and vectorization.
+
+This API aims to replace the existing implementation of Silk.NET.Maths.
+
+# Contributors
+- Maxine H (@uwx)
+
+# Current Status
+- [x] Proposed
+- [ ] Discussed with API Review Board (ARB)
+- [ ] Approved
+- [ ] Implemented
+
+# Design Decisions
+- Should replace the existing APIs without too much growing pains
+- Should try to roughly match System.Numerics APIs, with concessions for design oversights in those APIs
+- Should use vectorized implementations for methods when possible
+- Prefer generic specialization and extension methods, and keep the type of `T` as wide as possible in the struct itself
+- Primitives are read-only, `with` preferred for modification
+- Public fields instead of properties
+- Implement the following interfaces:
+    - IEquatable
+    - IEquatable to equivalent System.Numerics primitive, if available
+    - IReadOnlyList
+    - IUtf8SpanFormattable
+    - IUtf8SpanParsable
+- Provide explicit implementations of the following interfaces:
+    - INumberBase (except for methods which are relevant to vectors)
+- Provide implementations of the following interfaces, but as extension/static methods:
+    - IHyperbolicFunctions
+    - ITrigonometricFunctions
+    - ILogarithmicFunctions
+    - IExponentialFunctions
+    - IPowerFunctions
+    - IRootFunctions
+    (The reasoning for this is that the required interface types are more narrow than INumberBase)
```

Comment:
This is where part of the reasoning around vector types fundamentally need to be floating-point comes in. They frequently use and need so many concepts that only make sense for floating-point types.

By relegating these to extension methods, you are effectively taking core and frequent concepts of the type and moving them to be non standard functionality. That in turn limits the reusability with regards to other features, like generic math.

Reply:
These are static methods, not extension methods. They would be static either way, and would be placed into a non-generic type for convenience, so I don't personally see the problem here.

Reply:
When they’re not part of the interface, then you can’t allow implementers of IVector to guarantee access to the methods, for example

thus you can’t, in a generic context, ever use Length() or Sqrt() with a TVector, `where TVector : IVector<TVector, T>`

you could define your own generic extension method, but then that can’t specialize for arbitrary implementers of IVector, only the cases you explicitly codify (thus effectively limiting it to your own `Vector2/3/4D<T>`)

——————

For generic math to “work”, effectively every API you would want to use needs to be on the interface used for the constraints. You likewise don’t want people constraining to individual interfaces because changing that set of constraints in the future is breaking (say you realize you need RootFunctions in v2 but only constrained to TrigFunctions in v1)

There’s a lot of very subtle usability considerations here to make your types usable and properly versionable over time without causing binary breaks

Reply:
I think for usability purposes, we might need to live without float-requiring functions on generic TVector. My hope is that TVector won't be used commonly enough for this to be a serious problem, but this might need more discussion.

Reply:
I personally think that hinders one of the main use cases.

That is, you’ll basically force people who want to support float -or- double (arguably one of the most powerful and beneficial features here) to not use generic math and write 2 implementations

For example, a user couldn’t define and then successfully implement something like:
```
public static TVector AngleBetweenNormals<TVector, T>(TVector normal1, TVector normal2)
    where TVector : IVector<TVector, T>
    where T : IFloatingPointIeee754<T>
```

Because while the signature is fine, TVector then has no way to give access to optimized/specialized versions of the APIs needed for any TVector.

They are then stuck with defining separate implementations for Vector2D, Vector3D, and Vector4D without code sharing and basically pessimize the support for any other theoretical types that might implement IVector

Reply:
The int-based Vector case is most likely going to need to be possible. It's also required for backcompat with 2.X. I'm not really sure how to tackle this. 

I thought about moving the `T : IRootFunctions` check to be at runtime, but then there's no way to call T.Sqrt without reflection (as far as I can tell). By pure coincidence, the 2.X generated code hellscape could handle this use case just fine.

Reply:
Right, there is no good way to do this (and might not be any way to do it for several releases still)

This is one of the reasons I was pushing so heavily in Discord for considering two types (things that make sense for floats vs things that make sense for all types; and then maybe a common interface hierarchy), because you’re effectively taking a ton of mainstream scenarios and telling users “this is impossible for you to do”, in favor of making it convenient for a subset of functionality to also work on integers

this is then why we have INumberBase, INumber, IBinaryInteger, IFloatingPointIeee754, and various other interfaces. So users can pick a more concrete subset or a a mor generic subset

One could imagine IVectorBase and IVector, for example. The former being for all vectors and the latter being for floating point vectors (but with better names)

You could then define Vector2I<T> and Vector2F<T>. Users can then work with concrete types or TVector constrained to the relevant version if they want to support either.

There are a lot of possibilities to make it easy for users to succeed and still allow general/reusable algorithms for all the cases Silk cares about